File: python/Q90_subsets_ii.py
#
# @lc app=leetcode id=90 lang=python3
#
# [90] Subsets II
#


# @lc code=start
class Solution:
    def backtrack(self, index: int, subset: list[int]) -> None:
        if index == self.length:
            self.result.append(subset.copy())
            return

        subset.append(self.nums[index])
        self.backtrack(index + 1, subset)

        subset.pop()

        while index + 1 < self.length and self.nums[index] == self.nums[index + 1]:
            index += 1
        self.backtrack(index + 1, subset)

    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        self.nums: list[int] = nums
        self.nums.sort()

        self.length: int = len(nums)
        self.result: list[list[int]] = []

        self.backtrack(0, [])

        return self.result


# @lc code=end

File: rust/src/leetcode/Q90_subsets_ii.rs
/*
 * @lc app=leetcode id=90 lang=rust
 *
 * [90] Subsets II
 */

// @lc code=start
impl Solution {
    fn backtrack(
        nums: &Vec<i32>,
        current: &mut Vec<i32>,
        result: &mut Vec<Vec<i32>>,
        index: usize,
    ) {
        if index == nums.len() {
            result.push(current.clone());
            return;
        }

        current.push(nums[index]);
        Self::backtrack(nums, current, result, index + 1);

        current.pop();
        let mut new_index: usize = index;
        let candidate: i32 = nums[index];
        while new_index < nums.len() && nums[new_index] == candidate {
            new_index += 1;
        }
        Self::backtrack(nums, current, result, new_index);
    }

    pub fn subsets_with_dup(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut sorted_nums: Vec<i32> = nums;
        sorted_nums.sort();

        let mut result: Vec<Vec<i32>> = Vec::new();
        let mut current: Vec<i32> = Vec::new();

        Self::backtrack(&sorted_nums, &mut current, &mut result, 0);

        result
    }
}
// @lc code=end
